src/main/java/org/cru/godtools/domain/packages/PageStructure.java

```diff
@@ -143,10 +164,161 @@ public void mergeXmlContent(Document updatedPageLayout)
+			Node xmlContentNode = currentXmlContentNodeList.item(i);
+
+			if (xmlContentNode != null)
+			{
+				Element originalElement = (Element) xmlContentNode;
+				Element addElement = (Element) nodeToAdd;
+
+				boolean attrMatch = hasSameAttributes(originalElement, addElement);
+
+				//if the xml attributes are not the same but the node
+				//name is the same it's probably a new node
+				if (!attrMatch && originalElement.getNodeName().equals(addElement.getNodeName()))
+				{
+					String nodeToBeAddedString = XmlUtilities.xmlDocumentOrNodeToString(new DOMSource(nodeToAdd));
+
+					if (!currentXmlStringListOfElements.contains(nodeToBeAddedString))
+					{
+						Node targetNode = nodeToAdd.cloneNode(true);
+						Node nodeToBeImported = xmlContent.importNode(targetNode, true);
+						xmlContent.getDocumentElement().insertBefore(nodeToBeImported, xmlContentNode.getPreviousSibling());
+					}
+				}
+			}
+			else
+			{
+				Node targetNode = xmlContent.importNode(nodeToAdd, true);
+				currentXmlContentNodeList.item(1).appendChild(targetNode);
+			}
+		}
+	}
+
+	private boolean hasSameAttributes(Element oElement, Element aElement)
+	{
+		NamedNodeMap originalNamedNodeMap = oElement.getAttributes();
+		NamedNodeMap additionNamedNodeMap = aElement.getAttributes();
+
+		boolean attrMatch = true;
+
+		for (int n = 0; n < additionNamedNodeMap.getLength(); n++)
+        {
+            Attr a1 = (Attr) additionNamedNodeMap.item(n);
+            Attr o1 = (Attr) originalNamedNodeMap.item(n);
+
+            if (!o1.getName().equals(a1.getName()) || !o1.getValue().equals(a1.getValue()))
+            {
+                attrMatch = false;
+                break;
+            }
+        }
+		return attrMatch;
+	}
+
+	public void removeXmlContent(Document documentWithRemovableElements) throws XMLStreamException, IOException,
+			ParserConfigurationException,SAXException,TransformerException
+	{
+		XmlUtilities.verifyDifferentXml(xmlContent, documentWithRemovableElements);
+
+		ByteArrayOutputStream originalXmlByteArrayOStream = XmlDocumentStreamConverter.writeToByteArrayStream(xmlContent);
+		ByteArrayOutputStream removableElementsByteArrayOStream = XmlDocumentStreamConverter.writeToByteArrayStream(documentWithRemovableElements);
+
+		XMLEventReader originalXmlReader = getXmlEventReaderFromByteArray(originalXmlByteArrayOStream);
+		XMLEventReader removableElementsXmlReader = getXmlEventReaderFromByteArray(removableElementsByteArrayOStream);
+
+		ByteArrayOutputStream byteArrayForDocument = new ByteArrayOutputStream();
+
+		XMLEventWriter xmlEventWriter = XMLOutputFactory.newInstance().createXMLEventWriter(byteArrayForDocument);
+		List<String> xmlEventArrayList = Lists.newArrayList();
+
+		boolean isFound = true ;
+		String deleteTagName = "";
+
+		//I'm sure there's a better way, but for now I write the results
+		//to a list and in the next while loop, we check if the event (xml element)
+		//exists.
+		while (removableElementsXmlReader.hasNext())
```

Comment:
Yeah, this has been refactored a few times. (I'm still not satisfied)  I really don't mind doing it again until we have the best solution.